crates/typst-pdf/src/gradient.rs

```diff
@@ -155,22 +161,39 @@ fn shading_function(ctx: &mut PdfContext, gradient: &Gradient) -> Ref {
 
         // If the color space is HSL or HSV, and we cross the 0°/360° boundary,
         // we need to create two separate stops.
-        if gradient.space() == ColorSpace::Hsl || gradient.space() == ColorSpace::Hsv {
+        let hue_component_index = match gradient.space() {
+            ColorSpace::Hsl | ColorSpace::Hsv => Some(0),
+            ColorSpace::Oklch => Some(2),
+            _ => None,
+        };
+
+        if let Some(index) = hue_component_index {
             let t1 = first.1.get() as f32;
             let t2 = second.1.get() as f32;
-            let [h1, s1, x1, _] = first.0.to_space(gradient.space()).to_vec4();
-            let [h2, s2, x2, _] = second.0.to_space(gradient.space()).to_vec4();
+            let c1 = first.0.to_space(gradient.space()).to_vec4();
+            let c2 = second.0.to_space(gradient.space()).to_vec4();
+            let h1 = c1[index];
+            let h2 = c2[index];
 
             // Compute the intermediary stop at 360°.
-            if (h1 - h2).abs() > 180.0 {
-                let h1 = if h1 < h2 { h1 + 360.0 } else { h1 };
-                let h2 = if h2 < h1 { h2 + 360.0 } else { h2 };
+            if (c1[index] - c2[index]).abs() > 180.0 {
+                let (h1, h2) = if h1 < h2 { (h1 + 360.0, h2) } else { (h1, h2 + 360.0) };
 
                 // We compute where the crossing happens between zero and one
                 let t = (360.0 - h1) / (h2 - h1);
                 // We then map it back to the original range.
                 let t_prime = t * (t2 - t1) + t1;
 
+                // Interpolated components with fixed hue.
+                let components = |hue: f32| {
+                    let mut c = [0.0; 3];
+                    c[0] = c1[0] * (1.0 - t) + c2[0] * t;
+                    c[1] = c1[1] * (1.0 - t) + c2[1] * t;
+                    c[2] = c1[2] * (1.0 - t) + c2[2] * t;
+                    c[index] = hue;
```

Comment:
Personally I had written it differently but I guess this is okay, here is how I had written it:
```rs
col1.map(|v| {
    if i == hue_index {
        i += 1;
        1.0
    } else {
        i += 1;
        v * (1.0 - t) + col2[i - 1] * t
    }
})
```